### PR TITLE
Staging legacy fake cloud provider

### PR DESCRIPTION
**Pull request by mcrute**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Staging the fake Cloud Provider as part of [KEP 20190125-removing-in-tree-providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md). Staging repo setup here https://github.com/kubernetes/legacy-cloud-providers

Ref: #75910

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
/sig cloud-provider aws
/priority important-longterm
/cc @andrewsykim @cheftako

### CODE REVIEW

File: hack/.golint_failures [hunk 1]
```diff
@@ -50,7 +50,6 @@ pkg/apis/rbac/validation
 pkg/apis/storage
 pkg/apis/storage/v1
 pkg/apis/storage/v1beta1
-pkg/cloudprovider/providers/fake
```
Comment (andrewsykim) [inline hack/.golint_failures]: Might have been easier to just add the staging package in here so we can do the golint change in a separate PR :P. I think it's fine though since the golint fixes for fake seem fairly small

Comment (cheftako) [reply hack/.golint_failures]: +1

Comment (mcrute) [reply hack/.golint_failures]: That's a good point, I debated doing it as two PRs but opted for one because I felt bad about moving known untidy code to another package; boyscout rule and all that. If this is too much to review with them combined I can try to split it.

Comment (andrewsykim) [reply hack/.golint_failures]: `fake` is relatively small so I think it's still reasonable to review :)

**Comment by liggitt** [comment]
Won't we want/need a fake impl to remain in-tree for testing even after we drop legacy-cloud-providers out of tree? I wasn't expecting this one to move out of tree.

**Comment by andrewsykim** [comment]
> Moving this to legacy-cloud-providers means that anyone that wants to use the fake also has to take dependencies on all the cloudprovider libs legacy-cloud-provider requires

Good point, but there are also use-cases for importing fake cloud provider externally. Maybe a better home is `k8s.io/cloud-provider/fake`?

**Comment by liggitt** [comment]
> Good point, but there are also use-cases for importing fake cloud provider externally. Maybe a better home is `k8s.io/cloud-provider/fake`?

locating the fake in the same module as the interfaces it implements makes a lot of sense to me

**Comment by cheftako** [comment]
> > Moving this to legacy-cloud-providers means that anyone that wants to use the fake also has to take dependencies on all the cloudprovider libs legacy-cloud-provider requires
> 
> Good point, but there are also use-cases for importing fake cloud provider externally. Maybe a better home is `k8s.io/cloud-provider/fake`?

It would also be nice to just clearly show that cloud-provider (especially providers.go) is no longer part k8s/k8s.
Having everything cloud-provider related in staging is a significant accomplishment for the team.
It also provides a clean mechanism for dependency/injection.
I also believe that for both testing and example purposes, that fake should be structured the same way as real cloud-providers.

**Comment by mcrute** [comment]
> > Good point, but there are also use-cases for importing fake cloud provider externally. Maybe a better home is `k8s.io/cloud-provider/fake`?
> 
> locating the fake in the same module as the interfaces it implements makes a lot of sense to me

Makes sense. I've updated this PR to that end.

**Comment by mcrute** [comment]
Failing tests seem spurious...

/retest

**Comment by liggitt** [comment]
/approve
looks good to me. will leave lgtm to @cheftako or @andrewsykim

**Review by andrewsykim — APPROVED** [review]
/lgtm

Thanks @mcrute
